File: Solutions by Programming Task/Fibonacci sequence/Java/TailRecursive.java
public class Main {
    private static int Fib(int x) {
        record Impl() {
            public static int Func(int p, int n, int i) {
                return i == 0 ? n : Func(p + n, p, i - 1);
            }
        }
        return Impl.Func(1, 0, x);
    }
    public static void main(String[] args) {
        System.out.println(Fib(10));
    }
}

File: Solutions by Programming Task/Fibonacci sequence/Lua/tail_recursive.lua
local function fib(x)
    local function f(p, n, i)
        return i == 0 and n or f(p + n, p, i - 1)
    end

    return f(1, 0, x)
end

for x = 0, 10 do
    print(fib(x))
end
